Make GridLiner into an Artist

rcomer opened the pull request:
<!--

Thanks for contributing to cartopy!
Please use this template as a guide to streamline the pull request you are about to make.

Remember: it is significantly easier to merge small pull requests. Consider whether this pull
request could be broken into smaller parts before submitting.

-->



## Rationale

<!-- Please provide detail as to *why* you are making this change. -->
Thinking about https://github.com/SciTools/cartopy/issues/2246#issuecomment-1731581546, it is not intuitive to me that `Gridliner._draw_gridliner` should be called as part of finding the `GeoAxes`'s bbox, since this method does not seem to update the `GeoAxes`'s position.  I tried moving the call so it's only used at draw and not for finding the bbox, and the only test that failed uses `fig.tight_layout()` before the first draw.  Clearly `tight_layout` cannot create space for artists that do not yet exist.

I wonder if a better approach is to have the `GridLiner` itself be an artist, then it can interact with the Matplotlib layouts and draw machinery on its own terms.  This PR is my attempt at that.  I went looking at the container artists in Matplotlib's [offsetbox.py](https://github.com/matplotlib/matplotlib/blob/v3.8.0/lib/matplotlib/offsetbox.py) for ideas about how to do it.  The `LineCollection` and `Text` objects are no longer added to the axes directly, but are children of the `Gridliner`.

Fixes #2246.  Also closes #1658 as we can now remove the `GridLiner` using the standard `Artist.remove` method.  I can add tests for those but I wanted to get some feedback on this first in case I have gone completely off piste!

This also seems to help #2247 - I only see the busy kernel when I interact with the plot.

## Implications

<!-- If applicable, to the best of your knowledge, what are the implications of this change? -->
If user code is relying on the fact that the `LineCollection` and `Text` objects were added directly to the axes, I guess this will break that.

<!--
## Checklist

 * If you have not already done so, ensure you've read and signed the Contributor Licence Agreement (CLA).
   (See the [governance page](http://scitools.org.uk/governance.html) for the CLA and what to do with it).

 * If this is a new feature, please provide an example of its use in the description. We may want to make a
   follow-on pull request to put the example in the gallery!

 * Ensure there is a suitable item in the cartopy test suite for the change you are proposing.

-->

rcomer commented on lib/cartopy/mpl/gridliner.py:
```diff
@@ -436,10 +441,6 @@ def __init__(self, axes, crs, draw_labels=False, xlocator=None,
 
-        # Check visibility of labels at each draw event
-        # (or once drawn, only at resize event ?)
-        self.axes.figure.canvas.mpl_connect('draw_event', self._draw_event)
```
I don't understand why this was here.  I _think_ it's telling Matplotlib to call the method on each draw, but that was also happening by calling it within `GeoAxes.draw`.  So was this redundant or am I misunderstanding what this is for?

rcomer commented:
> Thinking about https://github.com/SciTools/cartopy/issues/2246#issuecomment-1731581546, it is not intuitive to me that Gridliner._draw_gridliner should be called as part of finding the GeoAxes's bbox, since this method does not seem to update the GeoAxes's position.

OK I've now realised that the axes' tightbbox incorporates all of its artists, so it does make sense to call `_draw_gridliner` there.  But that leads me to https://github.com/matplotlib/matplotlib/pull/26899

greglucas reviewed (commented):
This seems like a great approach.

We do have this open issue with another idea for gridlines as well if you wanted to look more into that approach: https://github.com/SciTools/cartopy/issues/1773
But I think it would be more effort than this and not exclusive of this either.

For a test you could add a mock draw method and count the number of draws you get for the gridlines.

rcomer commented on lib/cartopy/mpl/gridliner.py:
```diff
@@ -234,7 +232,14 @@ def __init__(self, axes, crs, draw_labels=False, xlocator=None,
+        super().__init__()
+        axes.add_artist(self)
+
+        # We do not want the labels clipped to axes.
+        self.set_clip_on(False)
```
Yes, the lines need clipping to the axes patch, which happens by default within `add_collection`
https://github.com/matplotlib/matplotlib/blob/ff552f45f2d30c76120fe0355306afac9a9dd2da/lib/matplotlib/axes/_base.py#L2255-L2256

I got a lot of image test failures before I realised that!

rcomer commented:
OK, I've added some tests.  Against `main` the new tests give me

```
FAILED lib/cartopy/tests/mpl/test_gridliner.py::test_gridliner_count_draws - AssertionError: Expected '_draw_gridliner' to have been called once. Called 2 times.
FAILED lib/cartopy/tests/mpl/test_gridliner.py::test_gridliner_remove - AttributeError: 'Gridliner' object has no attribute 'remove'
FAILED lib/cartopy/tests/mpl/test_gridliner.py::test_gridliner_save_tight_bbox - AttributeError: 'NoneType' object has no attribute 'dpi'
```

I also
* moved the `add_artist` call into the `ax.gridlines()` method as I think that's more consistent with mpl plotting methods
* got rid of the `ax._gridliners` attribute as these are now tracked through `ax.artists`
* updated a couple of geoaxes method docstrings to reflect the existing changes

rcomer commented:
Test failures all seem to be something to do with MODIS.

rcomer commented on lib/cartopy/tests/mpl/test_gridliner.py:
```diff
@@ -432,3 +437,38 @@ def test_gridliner_formatter_kwargs():
+        mocked.assert_called_once()
+
+
+def test_gridliner_remove():
```
I wanted a `check_figures_equal` for this one to show that adding and removing the gridliner gives the same as not adding it, but that doesn't seem to exist [yet] in pytest-mpl.  Another option would be an image test against an empty rectangle...

greglucas replied:
Can you test against an already existing image?
https://github.com/SciTools/cartopy/blob/main/lib/cartopy/tests/mpl/baseline_images/mpl/test_mpl_integration/simple_global.png

rcomer replied:
I can try - can you specify a different directory in the image test decorator?  Or should I just create a symlink from the gridliner baseline image directory to the target image?

greglucas replied:
I have no idea... I was just pointing you at an empty figure to test against :)

FWIW, I think your current version with the asserts is fine.

rcomer replied:
Hmmm.  I can report that the decorator takes a _baseline_dir_ keyword, but it doesn't appear to have any effect.

rcomer replied:
Ah, it's possible if we modify `conftest.py`.  I've put that in a separate commit in case we decide that's not a good idea...

lgolston commented:
> Test failures all seem to be something to do with MODIS.

This issue has been bumped recently at #geopython/OWSLib#769

dopplershift commented:
This does seem like a good approach.

rcomer commented on lib/cartopy/mpl/gridliner.py:
```diff
@@ -1239,6 +1233,26 @@ def _axes_domain(self, nx=None, ny=None):
+    def get_window_extent(self, renderer=None):
+        self._draw_gridliner(renderer=renderer)
+        visible_text_extents = [c.get_window_extent(renderer=renderer)
+                                for c in self.label_artists if c.get_visible()]
```
Although this works for all tested cases, it feels a bit wrong to be excluding the lines from the bbox calculation here.  Would it be better to define `get_tightbbox` directly as a union of the tightbboxes of the child artists?

greglucas replied:
I think that makes sense to get the union of all child artists as well because they are a part of this artist.

One of the other things I noticed with this when testing it out is if you pan up, the title moves locations based on the axis labels box on the y-axis, which doesn't happen in plain MPL. If the text for the axis extends beyond the spine, then the title gets pushed up with it. Then when the label disappears the title jumps back down again. So I'm wondering if this also needs some adjustment on that aspect as well... (example: https://github.com/SciTools/cartopy/issues/2246)

rcomer replied:
I can reproduce the oscillating title with `main` and with v0.22.  The culprit is this method:

https://github.com/SciTools/cartopy/blob/f2bb81d6bc3826a73ff4e0f67a28c0b04f3ab179/lib/cartopy/mpl/geoaxes.py#L533-L566

rcomer replied:
I added a commit that addresses the case above: we check whether the horizontal extents of the left/right labels overlap the horizontal extent of the axes.  If not, the labels do not affect the title position.  I'm not sure whether this is too simplistic or too projection-specific though.

We do not seem to have any existing test coverage for this title positioning method.  So I tried this example from when the code went in: https://github.com/SciTools/cartopy/pull/1117#issuecomment-445866100  The good news is that I get identical results with my branch and with `main`.  The bad news is I don't get what was shown in that comment.  Note the titles for `AlbersEqualArea` and `LambertConformal` are lower here than they were then.  A closer look at the `AlbersEqualArea` example shows that it has no "left", "right" or "top" labels, only "bottom" and "geo".  So it seems like "geo" needs to be incorporated into the method somehow.  I haven't really got my head around what the "geo" labels do though.

![gridliner_titling global branch](https://github.com/SciTools/cartopy/assets/10599679/407682fb-7373-498a-8c00-6805eabeffa2)

rcomer replied:
OK, I think I get it now: the left and right labels refer to anything on the vertical spines of the standard rectangular axes.  The "geo" labels are everywhere where the spine curves round (I think).  That example now gives me the below image, and I have pulled some of it out into a new image test.

![gridliner_titling global](https://github.com/SciTools/cartopy/assets/10599679/7bdf2735-e64a-4f51-a57e-22a7afec71a0)

greglucas commented on lib/cartopy/mpl/geoaxes.py:
```diff
@@ -532,21 +528,22 @@ def draw(self, renderer=None, **kwargs):
+        if not gridliners:
             return
 
         if self._autotitlepos is not None and not self._autotitlepos:
             return
```
Minor, but possibly put this check first to short-circuit out of it before iterating through artists?

greglucas commented on lib/cartopy/tests/mpl/test_gridliner.py:
```diff
@@ -242,8 +246,9 @@ def test_grid_labels_tight():
     for ax in fig.axes:
-        for gl in ax._gridliners:
-            assert hasattr(gl, '_drawn') and gl._drawn
+        for artist in ax.artists:
+            if isinstance(artist, Gridliner):
+                assert hasattr(artist, '_drawn') and artist._drawn
```
I realize this was not your doing, but this seems fragile if something changes types and we don't hit this code path. Could we change it to a count assertion `num_gridliners_drawn == expected_number` to make sure we are hitting the code path we expect?

greglucas reviewed (approved):
I think this is a good improvement.

rcomer commented:
Thanks @greglucas - I have addressed those two points and also updated a comment that I missed before.

greglucas commented:
All test failures are due to external WMTS server failures and are unrelated to this PR.
